torch/_subclasses/meta_utils.py

```diff
@@ -374,6 +374,36 @@ def all_dynamic_symbolic_context(t, source, shape_env, callback):
 
             return t_symbolic_context
 
+        def _nasty_nested_int_symbolicizing_hack():
+            # A nasty hack to deal with any NJT intermediates created during view replay.
+            # For these, we need to ensure that the corresponding nested ints are made
+            # symbolic.
+            #
+            # Example:
+            #     Consider a NJT1 -> Dense -> NJT2 -> Dense view. During view replay, the
+            #     Dense -> NJT2 part will construct an intermediate, symbolically-sized NJT
+            #     that is immediately deconstructed to return the final dense view. To
+            #     construct this intermediate properly, we need the associated nested int
+            #     to be symbolic. Here we accomplish that by patching the behavior of the
+            #     nested int constructor to wrap the result in an ephemeral SymInt.
+            from unittest.mock import patch
+
+            from torch._dynamo.source import EphemeralSource
+
+            old_get_nested_int = torch._C._get_nested_int
+
+            def symbolic_get_nested_int(_tensor_id_counter, coeff):
+                nested_int = old_get_nested_int(_tensor_id_counter, coeff)
+                sym_source = EphemeralSource("symbolic _get_nested_int()")
+                symbol = shape_env.create_symbol(nested_int, sym_source)
+                return shape_env.create_symintnode(
+                    symbol, hint=nested_int, source=sym_source
+                )
+
+            return patch(
+                "torch._C._get_nested_int", side_effect=symbolic_get_nested_int
+            )
```

Comment:
cc @ezyang / @soulitzer here's an annoying problem (explained in the example comment):
```
Consider a NJT1 -> Dense -> NJT2 -> Dense view. During view replay, the
Dense -> NJT2 part will construct an intermediate, symbolically-sized NJT
that is immediately deconstructed to return the final dense view. To
construct this intermediate properly, we need the associated nested int
to be symbolic. Here we accomplish that by patching the behavior of the
nested int constructor to wrap the result in an ephemeral SymInt.
```

So construction of new nested ints needs to be done symbolically during view replay. Can this go away once the nested int <-> offsets association is fully resolved?

Reply:
From discussion with Joel offline it sounds like we would eventually need some logic when replaying views to handle the case where we see a fake offsets (1) that we've never seen before, (2) has ephemeral source. And overall it would simplify the current situation here where we do patching.

Reply:
What happens when I do arithmetic on a SymInt with ephemeral source during view replay? This situation seems analogous.

Reply:
@soulitzer / @ezyang I removed the hack and added an xfailing test instead. AFAICT this will be solved with Jeffrey's changes to store the symbolic nested int onto the fake offsets tensor